tfhe/src/integer/server_key/radix_parallel/scalar_div_mod.rs

```diff
@@ -0,0 +1,481 @@
+    multiplier: u128,
+    // The shift that we might need
+    // to do post mutliplication to correct error
+    shift_post: u32,
+}
+
+impl ApproximatedMultiplier {
+    #[allow(non_snake_case)]
+    fn choose(divisor: u64, precision: u32, integer_bits: u32) -> Self {
+        // Keep the same name as in the paper
+        let N = integer_bits;
+
+        assert_ne!(divisor, 0);
+        assert!(precision >= 1 && precision <= N);
+
+        let d = u128::from(divisor);
+        let two_pow_n = 1u128 << N;
+
+        // Find l such that (l-1) < log2(d) <= l
+        let l = d.ilog2() + u32::from(!d.is_power_of_two());
+        let mut shift_post = l;
+
+        // The formula is m_low == 2**(N+l) / d
+        // however N + l may be == to 128, which means doing
+        // 2**(N + l) would not be valid, so the formula has been written in a
+        // way that does not overlflow.
+        //
+        // By writing m_low = (m_low - 2**N) + (2**N)
+        let mut m_low = (two_pow_n * ((1 << l) - d)) / d;
+        m_low += two_pow_n;
+
+        // The formula is (2**(N+l) + 2**(N + l - precision)) / d
+        // again 2**(N+l) could overflow
+        let mut m_high = ((two_pow_n * ((1 << l) - d)) + (1 << (N + l - precision))) / d;
+        m_high += two_pow_n;
+
+        assert!(m_low < m_high);
+
+        loop {
+            let m_low_i = m_low >> 1;
+            let m_high_i = m_high >> 1;
+
+            if m_low_i >= m_high_i || shift_post == 0 {
+                break;
+            }
+
+            m_low = m_low_i;
+            m_high = m_high_i;
+            shift_post -= 1;
+        }
+
+        Self {
+            multiplier: m_high,
+            shift_post,
+        }
+    }
+}
+
+impl ServerKey {
+    /// computes lhs * rhs on the full precision (num_block * 2) and returns
+    /// the most significant num blocks.
+    ///
+    /// In other words, it is like doing
+    /// fn mulhi (a: u32, b: u32) -> u32 {
+    ///    (a as 64 * b as u64) >> 32) as u32
+    /// }
+    fn scalar_mul_high(&self, lhs: &RadixCiphertext, rhs: u128) -> RadixCiphertext {
+        let mut result = lhs.clone();
+        self.extend_radix_with_trivial_zero_blocks_msb_assign(&mut result, lhs.blocks.len());
+        self.scalar_mul_assign_parallelized(&mut result, rhs);
+        self.trim_radix_blocks_lsb_assign(&mut result, lhs.blocks.len());
+        result
+    }
+
+    pub fn unchecked_scalar_div_parallelized<T>(
+        &self,
+        numerator: &RadixCiphertext,
+        divisor: T,
+    ) -> RadixCiphertext
+    where
+        u64: From<T>,
+        T: Numeric,
+    {
+        let divisor_bits = T::BITS as u32;
+        let numerator_bits = self.key.message_modulus.0.ilog2() * numerator.blocks.len() as u32;
+        assert!(
+            divisor_bits <= numerator_bits,
+            "The divisor has more bits than the numerator\
+                (numerator: {numerator_bits} bits, divisor: {divisor_bits} bits)"
+        );
```

Comment:
can we return Zero in that case ?

Reply:
No that would not be correct, its a check on the bit size of the type, not the value.

Maybe we can check on the value instead

Reply:
Changed the check, we now look at the value (its ilog2 + 1) if its bigger than the number of bits in the ciphertext
then the quotien is a trivial 0.

A corresponding test case has been added